File: 2188-minimized-maximum-of-products-distributed-to-any-store/minimized-maximum-of-products-distributed-to-any-store.py
class Solution:
    def minimizedMaximum(self, n: int, quantities: List[int]) -> int:
        low, high = 1, max(quantities)
        count = 0

        while low <= high:
            mid = low + (high - low) // 2
            need = sum(math.ceil(x / mid) for x in quantities)

            if need <= n:
                count = mid
                high = mid - 1
            else:
                low = mid + 1

        return count

File: 2188-minimized-maximum-of-products-distributed-to-any-store/minimized-maximum-of-products-distributed-to-any-store.js
/**
 * @param {number} n
 * @param {number[]} quantities
 * @return {number}
 */
var minimizedMaximum = function (n, quantities) {
    var low = 1;
    var high = Math.max(...quantities);
    var count = 0;

    const storesNeeded = (x) => {
        return quantities.reduce((sum, quantity) => sum + Math.ceil(quantity / x), 0);
    };

    while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        const need = storesNeeded(mid);

        if (need <= n){
            count = mid;
            high = mid - 1;
        }
        else{
            low = mid + 1;
        }
    }

    return count;
};